### DISCUSSION TITLE
Direnv performance improvement on shell startup time

### DISCUSSION THREAD

**Discussion by dagadbm** [asdf-community/asdf-direnv#57] [Show and tell]
Currently, using asdf-direnv and setting up a gloval .envrc file makes the shell unusably slow.

To combat this I did the following on my setup:

on my zshrc i just setup asdf regularly, and include the direnv hook which looks like this:

```
# asdf
. $HOME/.asdf/asdf.sh
fpath=(${ASDF_DIR}/completions $fpath)
autoload -Uz compinit
compinit
## asdf direnv setup
eval "$(asdf exec direnv hook zsh)"
```

afterwards i do create a global direnv file with the following contents, which are basically copy pasted from .asdf/asdf.sh

```
source $(asdf which direnv_use_asdf)

# Uncomment the following line to make direnv silent by default.
# export DIRENV_LOG_FORMAT=""

## Unset asdf shims to use asdf direnv properly
ASDF_BIN="$HOME/.asdf/bin"
ASDF_USER_SHIMS="$HOME/.asdf/shims"
[[ ":$PATH:" == *":${ASDF_BIN}:"* ]] && PATH="${PATH//$ASDF_BIN:/}"
[[ ":$PATH:" == *":${ASDF_USER_SHIMS}:"* ]] && PATH="${PATH//$ASDF_USER_SHIMS:/}"
export PATH="${ASDF_BIN}:$PATH"
```

What does this give me?

Basically I get all the benefits of using asdf globally with no cons regarding shell startup time by using asdf-direnv.

If for some reason I do indeed need some custom direnv solution which doesnt use my global asdf settings then ofc we go into direnv land.

This makes the shell startup time a little bit more bearable if you have a lot of asdf plugins installed.

If someone ends up on the same situation as me, this is how i solved it.

Aso, I do know you can use some direnv caching using nix, but I had never heard of nix before and I do not want to learn yet another new thing to do this so voila.

**Comment by michi-zuri** [comment]
A hybrid solution like you describe is a reasonable suggestion indeed if you need zippy startup times for the terminal. This may deserve an honorable mention in the readme.

**Comment by vic** [comment]
Yes, please consider improving our documentation.

**Comment by dagadbm** [comment]
I have just updated your plugin and this following setup works fine for me:

```
.zshrc
# asdf must be sourced before compinit
. $HOME/.asdf/asdf.sh
fpath=(${ASDF_DIR}/completions $fpath)
# Hook direnv into your shell.
eval "$(asdf exec direnv hook zsh)"
export DIRENV_LOG_FORMAT=
# ... afterwards i am sourcing oh my zsh (i know its nooby but i  dont know prezto or zinit)
# source $ZSH/oh-my-zsh.sh
```

on the direnvrc i only have this
```
source "$(asdf direnv hook asdf)"

# Fetch the nvim python provider if needed
use_nvim_python_provider() {
	pip show pynvim &> /dev/null || pip install pynvim &> /dev/null
}

# basic usage {{{

# .envrc
# use asdf
# layout python python3

# .tool-versions
# python 3.8.1
# nodejs 10.21.0
# }}}
```

**Comment by dagadbm** [comment]
but i also am using powerlevel10k prompt shell which might have probably fixed the start up time i had with the shell...

**Comment by brad-jones** [comment]
What would perhaps be neat is if the asdf direnv plugin actually generated your `.envrc` file for you instead of a `.tool-versions` file. I too came to this plugin in an attempt speed up asdf shims but then I just ended up with a slow terminal instead. What I have done is basically just used `asdf install` to install my tools & not bothered with the `asdf global/local` commands & thus no `.tool-versions` files.

And then with vanilla `direnv`  I have my `.envrc` files contain this sort of thing.

```sh
PATH_rm '$ASDF_INSTALLS_DIR/nodejs/*'
PATH_add "$ASDF_INSTALLS_DIR/nodejs/18.13.0/bin"
```

`direnv` it's self is basically instant now or at least no slower than say a single asdf shim might have been.

So yeah instead of `asdf direnv local [<tool> <version>]...` writing a `.tool-versions` file, get it to spit out a `.envrc` file or even a `.envrc.tool-versions` file which is sourceable from the main `.envrc` file. Basically pay the performance penalty at install time instead of every time I open a directory.

**Reply by brad-jones** [reply]
lol I just realized `asdf direnv envrc` is generating `.envrc` code & it gets cached into `~/.cache/asdf-direnv/env`.
I have basically manually written into my `.envrc` files what this plugin is generating hahaha

I wonder why it's so slow then? Feel like something about the way the caching works is not quite right.

Anyway again I wonder how hard it would be to refactor the `asdf direnv local` command to basically do what it does now plus perform similar logic to the current `asdf direnv envrc` command but spit the resulting `.envrc` out in the current directory. And then from that point in time it's all up to `direnv`.

I might hack around and see what I can come up with.

**Reply by jfly** [reply]
@brad-jones, you are right, the cached envrc file is *supposed* to make this quite fast. That has been my experience.

Perhaps there's a bug that's causing you to regenerate the cached file each time? It should be obvious from the output if it's getting reused or regenerated.

If it's getting regenerated, you can try turning on the debugging firehouse to figure out why: `export ASDF_DIRENV_DEBUG=true; direnv reload`.
